src/main/scala/com/fulcrumgenomics/bam/pileup/ByCoordinatePileupBuilder.scala

```diff
@@ -0,0 +1,182 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam.pileup
+
+import com.fulcrumgenomics.bam.api.SamOrder.Coordinate
+import com.fulcrumgenomics.bam.api.{SamIterator, SamOrder, SamRecord, SamSource}
+import com.fulcrumgenomics.bam.pileup.ByCoordinatePileupBuilder.DefaultInitialCacheSize
+import com.fulcrumgenomics.commons.CommonsDef._
+import com.fulcrumgenomics.coord.LocatableUtil.GenomicOrdering
+import com.fulcrumgenomics.fasta.SequenceDictionary
+import htsjdk.samtools.util.{Interval, Locatable}
+
+import java.io.Closeable
+import scala.collection.mutable
+
+/** A class to lazily pileup SAM records for coordinate-maintaining or coordinate-advancing loci.
+  *
+  * @param records a by-name iterator of SAM records that is assumed to be coordinate-sorted.
+  * @param dict the sequence dictionary associated with the reference sequences.
+  * @param minBaseQ the minimum base quality that a base must have to contribute to a pileup.
+  * @param minMapQ the minimum mapping quality a read must have to contribute to a pileup.
+  * @param mappedPairsOnly if true, only allow read-pairs with both reads mapped to contribute to a pileup.
+  * @param includeDuplicates if true, allow reads flagged as duplicates to contribute to a pileup.
+  * @param includeSecondaryAlignments if true, allow reads flagged as secondary alignments to contribute to a pileup.
+  * @param includeSupplementalAlignments if true, allow reads flagged as supplementary alignments to contribute to a pileup.
+  * @param excludeMapPositionOutsideFrInsert if true, exclude any record of an FR pair where the site requested is outside the insert.
+  * @param initialCacheSize the initial size for the internal SAM record cache, set this to your expected pileup depth.
+  */
+class ByCoordinatePileupBuilder private(
+  records: => SamIterator,
+  override val dict: SequenceDictionary,
+  override val minMapQ: Int                               = 20,
+  override val minBaseQ: Int                              = 20,
+  override val mappedPairsOnly: Boolean                   = true,
+  override val includeDuplicates: Boolean                 = false,
+  override val includeSecondaryAlignments: Boolean        = false,
+  override val includeSupplementalAlignments: Boolean     = false,
+  override val excludeMapPositionOutsideFrInsert: Boolean = true,
+  val initialCacheSize: Int                               = DefaultInitialCacheSize,
+) extends PileupBuilder(
```

Comment:
I explored a variety of APIs before deciding to use inheritance of the existing `PileupBuilder`.

No other API that I had designed made much sense given function, encapsulation of state, etc.